cmd/snap-discard-ns/snap-discard-ns.c

```diff
@@ -15,41 +15,190 @@
+	 * to die() */
+	int snap_lock_fd = sc_lock_snap(snap_instance_name);
+	debug("discarding mount namespaces of snap %s", snap_instance_name);
+
+	const char *ns_dir_path = "/run/snapd/ns";
+	int ns_dir_fd = open(ns_dir_path, O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (ns_dir_fd < 0) {
+		/* The directory may legitimately not exist if no snap has started to
+		 * prepare it. This is not an error condition. */
+		if (errno == ENOENT) {
+			return 0;
 		}
+		die("cannot open path %s", ns_dir_path);
+	}
+
+	/* Move to the namespace directory. This is used so that we don't need to
+	 * traverse the path over and over in our upcoming unmount2(2) calls. */
+	if (fchdir(ns_dir_fd) < 0) {
+		die("cannot move to directory %s", ns_dir_path);
 	}
 
+	/* Create shell patterns that describe the things we are interested in:
+	 *
+	 * Preserved mount namespaces to unmount and unlink:
+	 * - "$SNAP_INSTANCE_NAME.mnt"
+	 * - "$SNAP_INSTANCE_NAME.[0-9]+.mnt"
+	 *
+	 * Applied mount profiles to unlink:
+	 * - "snap.$SNAP_INSTANCE_NAME.fstab"
+	 * - "snap.$SNAP_INSTANCE_NAME.[0-9]+.fstab" */
+	char sys_fstab_pattern[PATH_MAX];
+	char usr_fstab_pattern[PATH_MAX];
+	char sys_mnt_pattern[PATH_MAX];
+	char usr_mnt_pattern[PATH_MAX];
+	sc_must_snprintf(sys_fstab_pattern, sizeof sys_fstab_pattern,
+			 "snap\\.%s\\.fstab", snap_instance_name);
+	sc_must_snprintf(usr_fstab_pattern, sizeof usr_fstab_pattern,
+			 "snap\\.%s\\.*\\.fstab", snap_instance_name);
+	sc_must_snprintf(sys_mnt_pattern, sizeof sys_mnt_pattern,
+			 "%s\\.mnt", snap_instance_name);
+	sc_must_snprintf(usr_mnt_pattern, sizeof usr_mnt_pattern,
+			 "%s\\.*\\.mnt", snap_instance_name);
+
+	DIR *ns_dir = fdopendir(ns_dir_fd);
+	if (ns_dir == NULL) {
+		die("cannot fdopendir");
+	}
+	/* ns_dir_fd is now owned by ns_dir and will not be closed. */
+
+	while (true) {
+		/* Reset errno ahead of any call to readdir to differentiate errors
+		 * from legitimate end of directory. */
+		errno = 0;
+		struct dirent *dent = readdir(ns_dir);
+		if (dent == NULL) {
+			if (errno != 0) {
+				die("cannot read next directory entry");
+			}
+			/* We've seen the whole directory. */
+			break;
+		}
+
+		/* We use dnet->d_name a lot so let's shorten it. */
+		const char *dname = dent->d_name;
+
+		/* Check the four patterns that we have against the name and set the
+		 * two should flags to decide further actions. Note that we always
+		 * unlink matching files so that is not reflected in the structure. */
+		bool should_unmount = false;
+		bool should_unlink = false;
+		struct variant {
+			const char *p;
+			bool unmount;
+		};
+		struct variant variants[4] = {
+			{.p = sys_mnt_pattern,.unmount = true},
+			{.p = usr_mnt_pattern,.unmount = true},
+			{.p = sys_fstab_pattern},
+			{.p = usr_fstab_pattern},
+		};
+		for (size_t i = 0; i < sizeof variants / sizeof *variants; ++i) {
+			struct variant *v = &variants[i];
+			debug("checking if %s matches %s", dname, v->p);
+			int match_result = fnmatch(v->p, dname, 0);
```

Comment:
The man of fnmatch says "FNM_NOMATCH if there is no match or another nonzero value if there is an error", so I think you should check against FNM_NOMATCH first before treating <0 as fatal.

Reply:
`FNM_NOMATCH` is 1 but I agree that this works by implementation detail, not by correctness. I'll changed that now.